Implement basic region-based state merging (RBSM 2/3)

futile opened the pull request:
This PR implements the basic version of a region-based state merging approach, behind the `-use-bounded-merge` flag. PR #691 adds additional functionality to this basic approach.

### Basic usage:
```c
klee_open_merge();

/* code containing branches etc. */

klee_close_merge();
```

Will lead to all states that forked from the state that executed the `klee_open_merge()` being merged in the `klee_close_merge()`. This allows for fine-grained regions to be specified for merging.

Tests for the implementation are also included in this PR, which give a few examples of how this technique can be used.

## Implementation Structure

The main part of the new functionality is implemented in the class `BoundedMergeHandler`. The Special Function Handler generates an instance of this class every time a state runs into a `klee_open_merge()` call.

This instance is appended to a `std::vector<klee::ref<BoundedMergeHandler>>` in the `ExecutionState` that passed the merge open point.  This stack is also copied during forks. We use a stack instead of a single instance to support nested merge regions.

Once a state runs into a `klee_close_merge()`, the Special Function Handler notifies the top `BoundedMergeHandler` in the state's stack, pauses the state from scheduling, and tries to merge it with all other states that already arrived at the same close merge point. This top instance is then popped from the stack, resulting in a decrease of the ref count of the `BoundedMergeHandler`.

Since the only references to this `BoundedMergeHandler` are in the stacks of the `ExecutionState`s currently in the merging region, once the ref count reaches zero, every state which ran into the same `klee_open_merge()` is now paused and waiting to be merged. The destructor of the `BoundedMergeHandler` then continues the scheduling of the corresponding paused states.

### Pausing and continuing states

For this purpose, we added the functionality to 'pause' a specific state from being scheduled again, but without terminating the state (commit e604ca1).

### Random Path Searcher

The `RandomPathSearcher` required additional changes to work with paused states (the `BumpMergingSearcher` also had problems with this, and would simply run into an endless loop if used together). In our version, we perform a simple backtracking step to find a non-paused state.

## Overhead

The biggest overhead introduced by this implementation is an additional vector of `Ref`s in each execution state. When state merging is not used, this vector will always be empty, and thus only take up the space required by an empty vector. In case state merging is used, the vector will contain an entry for each nested merging-region the state is in; this number should usually be rather low.

Additionally, some new member variables were added to the `Executor` class, but the additional space required should be negligible, especially if state merging is not used. However, the state-pausing functionality requires some bookkeeping inside `Executor::updateStates`, which is performed even if state merging is disabled (because the pausing/continuing of states is independent of state merging). This should still be extremely cheap, especially if there are no paused states (e.g., currently this means when state merging is not enabled).

Further the `RandomPathSearcher` has to check whether there are paused states, so the backtracking functionality can be used if required. This, too, should not make a visible performance difference, and also means that the `RandomPathSearcher` continues to work as before if state merging is turned off (the default).

ccadar commented on lib/Core/Searcher.cpp:
```diff
@@ -273,32 +321,129 @@ RandomPathSearcher::RandomPathSearcher(Executor &_executor)
+ExecutionState &RandomPathSearcher::selectStateStandard() {
+  unsigned flips=0, bits=0;
+  PTree::Node* n = executor.processTree->root;
+  BacktrackHelper btHelper;
```
This variable is unused.  Also, I would prefer to just rename the old `selectState` to `selectStateStandard` w/o reformatting it.

ccadar commented on lib/Core/Searcher.cpp:
```diff
@@ -273,32 +321,129 @@ RandomPathSearcher::RandomPathSearcher(Executor &_executor)
+  }
+  return *n->data;
+}
+ExecutionState &RandomPathSearcher::selectStateStandard() {
```
Add an empty line here

ccadar commented on lib/Core/Searcher.h:
```diff
@@ -165,6 +165,26 @@ namespace klee {
 
   class RandomPathSearcher : public Searcher {
     Executor &executor;
+    class BacktrackHelper {
```
Please add some comments for this helper class, e.g., by describing the `visited` array and the directions.

ccadar commented on lib/Core/Searcher.h:
```diff
@@ -165,6 +165,26 @@ namespace klee {
+        void go(direction dir);
+    };
+
+    std::set<ExecutionState *> ignoreStates;
```
Please add a comment about the role of `ignoreStates`

ccadar commented on lib/Core/Executor.h:
```diff
@@ -147,6 +147,11 @@ class Executor : public Interpreter {
 
+  /// Used to track states that are only removed from the searcher
+  std::vector<ExecutionState *> pausedStates;
+  /// Used to track states that are only added to the searcher
```
Can you expand on this comment? What does "only added to the searcher" means, exactly?

ccadar commented on include/klee/BoundedMergeHandler.h:
```diff
@@ -0,0 +1,42 @@
+
+public:
+
+  void addClosedState(ExecutionState *es, llvm::Instruction *mp);
```
Please add some comments for these functions

ccadar commented on lib/Core/BoundedMergeHandler.cpp:
```diff
@@ -0,0 +1,72 @@
+#include "klee/BoundedMergeHandler.h"
```
Please add the standard header with the KLEE license to this file.

futile replied:
Added the standard header with the KLEE license.

ccadar commented on include/klee/BoundedMergeHandler.h:
```diff
@@ -0,0 +1,42 @@
+#pragma once
```
Your comments added to this PR are excellent.  Please add them here, in doxygen format. 
Also, please add the standard header with the KLEE license to this file.

ccadar replied:
Also add the standard KLEE licence header to this file

futile replied:
Added the standard header with the KLEE license and a doxygen comment at the top explaining the functionality (taken from this PR).

ccadar commented:
Hi @futile , thanks again for these patches.  This looks like a great feature overall, which would be nice to add to KLEE.  But since it is a big change, it will require some iterations and more testing.  I hope you won't find the process too exhaustive!  I added a first round of reviews.  You will unfortunately also need to rebase and resolve a few conflicts, because your patches got caught in the middle of our removal of LLVM 2.9 support.

futile commented:
> Hi @futile , thanks again for these patches. This looks like a great feature overall, which would be nice to add to KLEE. But since it is a big change, it will require some iterations and more testing. I hope you won't find the process too exhaustive! I added a first round of reviews. You will unfortunately also need to rebase and resolve a few conflicts, because your patches got caught in the middle of our removal of LLVM 2.9 support.

Hi again, and thank you for taking your time to review our pull request! Sure, I am fine with some more iterations and more testing. The first round of reviews should be addressed, with all relevant comments/license headers added (other things too). We will rebase against the current master in the next few days. If you have concrete ideas of what kind of testing would be nice to have (you mentioned a specific testsuite before, I believe), we can also run that. In the meantime we are also looking into possible testing scenarios, in addition to what is already provided by the unit tests.

ccadar commented:
Hi @futile , I have finally managed to review this PR, sorry for the long delay!  The new merging support is indeed more flexible than the old one.  At the same time, it's nice that you simply reused the `merge()` function, which is quite complex.

I think after your latest changes, we should be able to merge your changes relatively soon, but here are a few more requests:

(1) Can you remove all the `RandomPathSearcher` commits and move them to a different PR?  This will speed up the addition of the core merging functionality, especially since I have more comments about the `RandomPathSearcher` changes. (Of course, the random search won't work anymore with merging, and you'll also have to remove those tests exercising it).

(2) What's the role of `BoundedMergingSearcher`?   Unless I'm missing something, it doesn't seem to be needed.

(3) I have a few more comments about specific parts of the code, which I'll add shortly.

(4) The code is often not indented properly, one option would be to use `clang-format` to assist with this.  I won't add comments about this now, as it can be easily resolved at the end.

Thanks!

ccadar commented on include/klee/BoundedMergeHandler.h:
```diff
@@ -0,0 +1,101 @@
+ * then continues the scheduling of the corresponding paused states.
+*/
+
+#pragma once
```
I would prefer to use an `ifndef...` here instead of `pragma one`, for consistency with all other header files.

ccadar commented on include/klee/BoundedMergeHandler.h:
```diff
@@ -0,0 +1,101 @@
+
+  /// @brief Called when a state runs into a 'klee_close_merge()' call
+  void addClosedState(ExecutionState *es, llvm::Instruction *mp);
+  /// @brief True, if any states have run into 'klee_close_merge()' and have
```
Can you add empty lines before these definitions, as you do above?

ccadar commented on include/klee/Internal/Module/KModule.h:
```diff
@@ -82,6 +82,10 @@ namespace klee {
     llvm::Module *module;
     llvm::DataLayout *targetData;
+    
+    // Some useful functions to know the address of
```
These don't seem necessary

ccadar commented on lib/Core/Executor.h:
```diff
@@ -133,6 +136,8 @@ class Executor : public Interpreter {
   std::vector<TimerInfo*> timers;
   PTree *processTree;
 
+  std::vector<BoundedMergeHandler *> mergeGroups;
```
Please add a comment about this data structure

ccadar commented on lib/Core/SpecialFunctionHandler.cpp:
```diff
@@ -324,6 +327,40 @@ void SpecialFunctionHandler::handleReportError(ExecutionState &state,
+    KInstruction *target,
+    std::vector<ref<Expr> > &arguments) {
+  if (!UseBoundedMerge) {
+    return;
```
I would add a `klee_warning` here saying that `klee_open_merge` is ignored as `use-bounded-merge` is not set.

ccadar commented on lib/Core/SpecialFunctionHandler.cpp:
```diff
@@ -324,6 +327,40 @@ void SpecialFunctionHandler::handleReportError(ExecutionState &state,
+    KInstruction *target,
+    std::vector<ref<Expr> > &arguments) {
+  if (!UseBoundedMerge) {
+    return;
```
Same here

ccadar commented on test/BoundedMerge/Behaviour/all_searcher/infinite_merge.c.ignore:
```diff
@@ -0,0 +1,44 @@
+// RUN: %S/setup_test.sh "%llvmgcc" %s %t "--search=nurs:covnew --use-batching-search" | FileCheck %s
```
Instead of adding this as an inactive test, just remove the unsupported parts.

ccadar commented:
I forgot:

(5) Regarding the tests you added, I would get rid of `setup_test.sh`, it's much better to inline it in each test.  Also, I don't think you need all that directory hierarchy, I would put everything in a "Merging" or "Merge" directory, properly named.  But if you prefer a hierarchy, I can accept this.

(6) Why call the option `use-bounded-merge` (and similarly for the code)?  Your PR refers to this as "region-based merging" which seems a better name.  However, I think "merge" might be enough.

(7) I'm wondering, have you considered using labels in `klee_open_merge` and `klee_close_merge` to ensure the merging is done as intended by the user?

futile commented:
Hello @ccadar,

Thank you for your further feedback!

(1) We removed the Random-path searcher modifications from this PR and will open up a new one for them.

(2) You are right, the `BoundedMergingSearcher` only becomes relevant in the PR that adds support for incomplete merging search strategies (RBSM 3/3), and it seems like a small part of that slipped into this PR (which we have now removed).

(3) Addressed the code-specific comments.

(4) We weren't sure whether to run our code/files through clang format or not, because it seems to be done in some places but not everywhere. But we can run clang format or otherwise reformat our code at any time, just let us know (as one of the last steps before merging sounds good).

(5) Got rid of `setup_test.sh` and flattened the test hierarchy (only a single directory called 'Merging' now).

(6) Renamed the option to `-use-merge`.

(7) Yes, we have considered a few variants of that, but in our opinion it brings only a small benefit while adding additional complexity and burden on the user, which is why we didn't include a variant of that in our implementation.

Thanks!

MartinNowack commented on lib/Basic/CmdLineOptions.cpp:
```diff
@@ -96,14 +105,6 @@ void KCommandLine::HideUnrelatedOptions(cl::OptionCategory &Category) {
 
-void KCommandLine::HideUnrelatedOptions(
-    ArrayRef<const cl::OptionCategory *> Categories) {
-  for (ArrayRef<const cl::OptionCategory *>::iterator i = Categories.begin(),
```
Why do remove this part?

futile replied:
This was accidentally removed during the last rebase, we added it back in.

MartinNowack commented on lib/Core/ExecutionState.cpp:
```diff
@@ -17,6 +17,12 @@
 #include "klee/Expr.h"
 
 #include "Memory.h"
+
```
None of this added includes should be needed? Am I mistaken?

MartinNowack commented on lib/Core/Executor.h:
```diff
@@ -78,13 +79,15 @@ namespace klee {
   /// removedStates, and haltExecution, among others.
 
 class Executor : public Interpreter {
+  friend class MergingSearcher;
```
Duplicate line: already exists two later

futile replied:
Fixed.

MartinNowack commented on test/Merging/abort.c:
```diff
@@ -0,0 +1,714 @@
+// RUN: %llvmgcc -emit-llvm -g -c -o %t.bc %s
```
What is the idea of this test case?
I see 700 new lines of code, but I have no glue what they are for.
Please add some description and add checks which you would like to use to validate your approach.

futile replied:
The purpose of this test was to check that there would be no segfaults/crashes in case klee runs out of time and halts execution. However, if necessary, this can also be done more simply (e.g., a loop, ..). For now, we removed this version of the test.

MartinNowack commented on include/klee/MergeHandler.h:
```diff
@@ -0,0 +1,106 @@
+#ifndef KLEE_MERGEHANDLER_H
+#define KLEE_MERGEHANDLER_H
+
+#include <set>
```
`set` is not needed.

futile replied:
Fixed.

MartinNowack commented on include/klee/MergeHandler.h:
```diff
@@ -0,0 +1,106 @@
+//===-- dMergeHandler.h --------------------------------------------*- C++ -*-===//
```
typo `d`

futile replied:
Fixed.

ccadar commented:
@futile , sounds great, thanks for your answers and recent changes!   I'll do another full pass over this soonish, but in the meantime please address @MartinNowack's code comments.

futile commented:
Addressed the code comments.

MartinNowack commented on include/klee/MergeHandler.h:
```diff
@@ -0,0 +1,105 @@
+//===----------------------------------------------------------------------===//
+
+/** 
+ * @file dMergeHandler.h
```
one "d" too much

MartinNowack commented on lib/Core/Executor.cpp:
```diff
@@ -2454,6 +2454,9 @@ void Executor::executeInstruction(ExecutionState &state, KInstruction *ki) {
 void Executor::updateStates(ExecutionState *current) {
   if (searcher) {
     searcher->update(current, addedStates, removedStates);
+    searcher->update(0, continuedStates, pausedStates);
```
If you can, it would be great if you can start using `c++11`.
* `0`-> `nullptr`

MartinNowack commented on lib/Core/Executor.cpp:
```diff
@@ -2739,6 +2742,30 @@ std::string Executor::getAddressInfo(ExecutionState &state,
 }
 
+void Executor::pauseState(ExecutionState &state){
+  std::vector<ExecutionState *>::iterator it = std::find(continuedStates.begin(), continuedStates.end(), &state);
```
`auto` instead of `std::vector<ExecutionState *>::iterator`

MartinNowack commented on lib/Core/Executor.cpp:
```diff
@@ -2739,6 +2742,30 @@ std::string Executor::getAddressInfo(ExecutionState &state,
+}
+
+void Executor::continueState(ExecutionState &state){
+  std::vector<ExecutionState *>::iterator it = std::find(pausedStates.begin(), pausedStates.end(), &state);
```
`auto` again

MartinNowack commented on lib/Core/MergeHandler.cpp:
```diff
@@ -0,0 +1,75 @@
+
+void MergeHandler::addClosedState(ExecutionState *es,
+                                         llvm::Instruction *mp) {
+  std::map<llvm::Instruction *, std::vector<ExecutionState *> >::iterator
```
`auto`

MartinNowack commented on lib/Core/MergeHandler.cpp:
```diff
@@ -0,0 +1,75 @@
+  } else {
+    // Otherwise try to merge with any state in the map element for this
+    // instruction
+    std::vector<ExecutionState *> &cpv = closePoint->second;
```
`auto &`

MartinNowack commented on lib/Core/MergeHandler.cpp:
```diff
@@ -0,0 +1,75 @@
+    std::vector<ExecutionState *> &cpv = closePoint->second;
+    bool mergedSuccessful = false;
+
+    for (std::vector<ExecutionState *>::iterator it = cpv.begin(),
```
`for (auto &mState: cpv) {`

MartinNowack commented on lib/Core/MergeHandler.cpp:
```diff
@@ -0,0 +1,75 @@
+}
+
+void MergeHandler::releaseStates() {
+  for (std::map<llvm::Instruction *, std::vector<ExecutionState *> >::iterator
```
simplify `for`loops as well

MartinNowack commented on lib/Core/SpecialFunctionHandler.cpp:
```diff
@@ -324,6 +327,42 @@ void SpecialFunctionHandler::handleReportError(ExecutionState &state,
+    llvm::errs() << "close merge: " << &state << " at " << i << '\n';
+
+  if (state.openMergeStack.empty()) {
+    llvm::errs() << &state << " ran into a close at " << i
```
use `klee_warning` here, it's independent of your logging

futile commented:
Hi,

we are currently busy updating our infrastructure to the CMake-based build system, but this hopefully won't take too long.

As far as this PR goes, is it correct to assume that discussion has - mostly - come to a close? The remaining points seem to be rather small code changes (+ final formatting), which should be done easily once we finish updating to the new build system. Or are there major points remaining that I missed?

ccadar commented:
Hi @futile , yes, thanks for your patience.  I'd still like to to a final pass over these changes, but I'll try to do this over the weekend.

ccadar commented on lib/Core/MergeHandler.cpp:
```diff
@@ -0,0 +1,75 @@
+      }
+    }
+    if (!mergedSuccessful) {
+      cpv.push_back(es);
```
It would be useful to have a test case for the case where the merging fails

ccadar commented on test/Merging/batching_break.c:
```diff
@@ -0,0 +1,42 @@
+
+// CHECK: open merge:
+// CHECK: close merge:
+// CHECK: KLEE: done: generated tests = 3{{$}}
```
You don't use `{{$}}`in any other tests.  I think you should always use it, or alternatively pass `--match-full-lines` to `FileCheck`

ccadar commented on test/Merging/indirect_value.c:
```diff
@@ -0,0 +1,33 @@
+#include <klee/klee.h>
+
+int main(int argc, char** argv) {
+  (void)argc; (void)argv;
```
Can you remove this line?  You don't have it in other test files and I don't think it triggers a warning without it, does it?

ccadar commented on test/Merging/nested_merge.c:
```diff
@@ -0,0 +1,48 @@
+// RUN: klee --output-dir=%t.klee-out --use-merge --debug-log-merge --search=nurs:covnew %t.bc 2>&1 | FileCheck %s
+
+// CHECK: open merge:
+// 5 close merges
```
Is there a way in `FileCheck` to specify the number of matches?  Maybe not.

ccadar commented:
Hi @futile , I think this looks great, my only other non-trivial request is to add an extra test to deal with the case where merging two states fails (I left a comment in the code).  I think this is important, as it would also document how the current merging interface works when this happens.  An actual comment discussing this case would be useful too.  Thanks again, I think we're very close to merge this.

kren1 commented:
As a minor polish point, I think it would be nice to change the default search strategy if -use-merge is used and -search isn't explicitly specified. Currently you just get an error, when just using -use-merge, which is not very friendly

ccadar commented on lib/Core/SpecialFunctionHandler.cpp:
```diff
@@ -355,8 +356,9 @@ void SpecialFunctionHandler::handleCloseMerge(ExecutionState &state,
   if (state.openMergeStack.empty()) {
-    klee_warning(std::to_string(&state) + " ran into a close at " + std::to_string(i)
-                 + " without a preceding open\n");
+    std::ostringstream warning;
```
This is minor, but I actually preferred to simply use `klee_warning` as you did before, instead of using an extra `std::ostringstream`

ccadar commented on lib/Core/UserSearcher.cpp:
```diff
@@ -88,8 +88,13 @@ Searcher *klee::constructUserSearcher(Executor &executor) {
-    CoreSearch.push_back(Searcher::NURS_CovNew);
+    if (UseMerge){
+      CoreSearch.push_back(Searcher::NURS_CovNew);
+      klee_warning("UseMerge enabled. Using NURS_CovNew as default searcher.");
```
Say `--use-merge` instead

ccadar commented:
@futile @corrodedHash many thanks for the additional changes, I believe we're ready to merge now!  Can you just rebase your changes and squash them into a smaller number of commits?  Thanks!

ccadar commented:
@futile , @corrodedHash , just a ping about rebasing this so that we can merge it.  We have another larger PR in the works, and I think it would be better to see this added to the mainline first.  Thanks!

futile commented:
@ccadar Many thanks! The Changes have been squashed and rebased. Looking at the output of the CI run, the failure seems to be unrelated:

```
$ if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then brew update && brew tap andreamattavelli/klee; fi
error: RPC failed; curl 56 SSLRead() return error -36
fatal: The remote end hung up unexpectedly
fatal: early EOF
fatal: index-pack failed
Error: Fetching /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core failed!
```

Thanks again for the continuous feedback to get this merged!

ccadar commented:
I think this now looks great and I plan to merge this shortly.  @MartinNowack , I think this also addresses your comments, but if there are any remaining smaller issues, I'd prefer to address them separately; I'll wait until tomorrow for any last-minute checks from your side, and then merge.  Thanks again!

andreamattavelli commented on lib/Basic/CmdLineOptions.cpp:
```diff
@@ -70,6 +70,15 @@ cl::opt<bool>
                           cl::desc("Optimize constant divides into add/shift/multiplies before passing to core SMT solver (default=off)"),
                           cl::init(false));
+llvm::cl::opt<bool>
+    UseMerge("use-merge",
```
`CmdLineOptions` is for shared KLEE/Kleaver options, and thus is used for solver-related command-line options.
I guess the best place where to put these declarations is `MergeHandler`.

ccadar replied:
This is a good catch, thanks, @andreamattavelli .  
@futile , @corrodedHash , if you can fix this quickly, that would be great.  Otherwise, I'd still prefer to merge this PR sometime tomorrow and we can fix this and other smaller issues separately.

futile replied:
Good point! We can fix this by Monday (or Tuesday at the latest), would that be fast enough?

ccadar replied:
That would work well, we can wait until Monday or Tuesday, thanks!

ccadar commented:
@corrodedHash @futile Thanks for the last commit, it solves that last issue.  Can you quickly squash it in the previous commit though?  It doesn't make sense to have it separately.

futile commented:
Moved the command line argument and squashed the commits.

ccadar commented:
Perfect, let's merge!  @andreamattavelli , @MartinNowack , I assume you're happy with this too.

ccadar commented:
@andreamattavelli , @MartinNowack , I'll merge these changes tomorrow, in case you see anything else blocking this, but I hope not, as we're approaching 70 exchanged messages on this!
@futile @corrodedHash , thanks again for your patience and contribution!

MartinNowack commented:
@ccadar From my side it's fine. Please go ahead.

ccadar commented:
Thanks, @MartinNowack .   @andreamattavelli ?

ccadar commented:
Merged!  Glad to see this in, thanks again, @futile , @corrodedHash .

futile commented:
Wooho! Thanks @ccadar @MartinNowack and @andreamattavelli for the continuous feedback and effort to get this merged! Great work @corrodedHash!
